### ISSUE TITLE
Change the manager classes (e.g. `ApplicationsManager`, `EnginesManager`) to provide synchronous methods

### ISSUE THREAD

**Issue by madoar** [PhoenicisOrg/phoenicis#1661]
Currently our manager classes only provide asynchronous methods taking a callback method as a parameter.
This callback method is then called after the manager method has finished its task.

Often it's needed to chain multiple manager methods behind each, where the current method invocation uses the result of the last one.
When using the asynchronous methods this can only be done via nesting, which can lead to deep nesting scopes.

I suggest to add a second set of methods, which are synchronous in behavior and directly return the result of the method.

**Comment by plata** [comment]
Other possible solution: only synchronous and the caller must take care that they are executed asynchronously.

**Comment by madoar** [comment]
That was my long term goal.
The issue here is, that the asynchronous methods are currently used by the JavaFX and CLI applications, and if I'm not wrong also the current web backend.

Therefore if we remove the asynchronous methods we not only need to change the JavaFX application but also the CLI and likely also the web backend to retain their functionality.

**Comment by qparis** [comment]
The real advantage I see about adynchroneous method, is that (if I’m not wrong) they are using phoenicis-multithreading which is a good pattern when it comes to dealing with multi threading issues.

Maybe it is overkill though

**Comment by madoar** [comment]
Without looking at `phoenicis-multithreading` in-depth, the multi-threading issue sounds like an use-case for the [Executors](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Executors.html) class to me.

In addition I think it is a better choice to let the developer decide for each use-case, whether he wants to execute the manager code in a synchronous or an asynchronous way.

**Comment by madoar** [comment]
I understand the benefits of an asynchronous approach.
The issue is, that it has some problems:
- you don't have control when the code is executed
- you don't know when the code has been finished, i.e. there is no return value
- chains of multiple asynchronous methods, each with its own callback method parameter, result in deep nesting levels.
   If I understand the code correctly it's also the case, that each nested callback method is called as its own task and maybe even thread.
   This is often overkill, because often all calls should be called in a synchronous manner in a single separate thread.

We also shouldn't forget, that my suggestion in this issue was only to add an additional set of synchronous methods and not to remove the asynchronous methods.
This is still a topic for future discussion :)

**Comment by qparis** [comment]
Yep, I agree to add synchroneous method. However, how I'd like to think about a general and maybe automated pattern to manage asynchroneous call cleanly in POLv5

**Comment by plata** [comment]
I fully agree to @madoar regarding the nested asynchronous calls.

Regarding a general solution: We could move the responsibility for asynchronous calls to the caller and only have a helper class providing the functionality.
Something like
```Java
Helper.runAsynchronous(repositoryManager::fetch, repositoryDTO -> { ... }
```

**Comment by qparis** [comment]
It is only because repositoryManager is doing different kind of tasks which are sent to differents thread pool

**Comment by qparis** [comment]
Syncing about it, I now see the problem of this proposal.
If you create a synchronous method, you will no longer be able to separate the sub calls and assign them to the right pools

**Comment by plata** [comment]
That's correct. Anyways: Let's discuss this later. We should concentrate on alpha-2 first.

**Comment by madoar** [comment]
@qparis this depends on how you define the `Helper.runAsynchronous(...)` method.
If you add an additional `threadPool` variable as a parameter to the method you don't have the problem, because you would still be able to create a nested call structure.

An additional solution idea would be a fluent API as a wrapper around the asynchronous calls.
This fluent API could then be used like the following:

```java
FluentManager
   .fetchRepository(repository -> {...})
   // fetch the engines from the previously fetched repository
   .fetchEngines(engines -> {...})
   // fetch the engine version belonging to the previously fetched engines
   .fetchVersions(versions -> {...})
   .execute();
```

This idea isn't yet completely thought-out.

**Comment by qparis** [comment]
Yes it should be something like this. However, it should not be a static class. Let's talk about it after alpha-2

**Comment by madoar** [comment]
I have a new idea how to solve this:
Can we maybe change our method signatures to something like:

```java
Future<Repository> fetchRepository();
Future<List<Engine>> fetchEngines(Repository repository);
```

The returned `Future<A>` object would then need to provide two methods:
- a `waitAndGet()` method, which waits until the corresponding value has been fetched and then returns it (synchronous)
- an `execute(callback)` method, which executes the given callback when the value has been fetched in an asynchronous manner

What do you think?

**Comment by qparis** [comment]
If you can chain the future inside several pools, this is the right way of doing this yes

**Comment by madoar** [comment]
This should be possible by providing a pool configuration to the `execute` method, right?
